drivers/spi/spi_mcux_lpspi.c

```diff
@@ -585,14 +615,129 @@ static int spi_mcux_init(const struct device *dev)
 	return 0;
 }
 
+#ifdef CONFIG_SPI_RTIO
+static void spi_mcux_iodev_complete(const struct device *dev, int status);
+static void spi_mcux_iodev_next(const struct device *dev, bool completion);
+
+static void spi_mcux_iodev_start(const struct device *dev)
+{
+	struct spi_mcux_data *data = dev->data;
+	struct rtio_sqe *sqe = &data->txn_curr->sqe;
+	struct spi_dt_spec *dt_spec = sqe->iodev->data;
+	struct spi_buf tx_buffer;
+	struct spi_buf rx_buffer;
+	struct spi_buf_set tx;
+	struct spi_buf_set rx;
+	int ret = 0;
+
+	switch (sqe->op) {
+		case RTIO_OP_TX:
+		case RTIO_OP_RX:
+		{
+			tx_buffer.buf = 0;
+			tx_buffer.len = sqe->buf_len;
+			rx_buffer.buf = sqe->buf;
+			rx_buffer.len = sqe->buf_len;
+
+			} break;
+		case RTIO_OP_TXRX:
+		{
+			tx_buffer.buf = sqe->tx_buf;
+			tx_buffer.len = sqe->txrx_buf_len;
+			rx_buffer.buf = sqe->rx_buf;
+			rx_buffer.len = sqe->txrx_buf_len;
+		} break;
+		default:
+		{
+			LOG_ERR("Invalid op code %d for submission %p\n", sqe->op, (void *)sqe);
+			struct rtio_iodev_sqe *txn_head = data->txn_head;
+
+			spi_mcux_iodev_next(dev, true);
+			rtio_iodev_sqe_err(txn_head, -EINVAL);
+			ret = 0;
+		}
+	}
+
+	tx.buffers = &tx_buffer;
+	tx.count = 1;
+	rx.buffers = &rx_buffer;
+	rx.count = 1;
+
+	ret = transceive(dev, &dt_spec->config, &tx, &rx, 0, 0, 0);
+	if (ret == 0) {
+		spi_mcux_iodev_complete(dev, 0);
+	}
+}
+
+static void spi_mcux_iodev_next(const struct device *dev, bool completion)
+{
+	struct spi_mcux_data *data = dev->data;
+	struct rtio_mpsc_node *next = rtio_mpsc_pop(&data->iodev.iodev_sq);
+
+	if (next != NULL) {
```

Comment:
next needs to be serialized in a spin lock as it is in the sam driver

It also needs to verify there's no ongoing current work before popping and starting the next thing on the queue.

Reply:
I wondered about that, but when the LPSPI driver reaches the transceive function, would this not stop the race condition?
https://github.com/zephyrproject-rtos/zephyr/blob/9686f6e000c746538ec7ab78f6d8abf299d24e95/drivers/spi/spi_mcux_lpspi.c#L488C1-L488C1

Reply:
> I wondered about that, but when the LPSPI driver reaches the transceive function, would this not stop the race condition?

looks like it would not, data is already changed before getting to this

Reply:
Thank you both, I have fixed this and there should no longer be a race condition.